### ISSUE TITLE
Make data in layer control editable by using copy-on-write

### ISSUE THREAD

**Issue by jipclaassens** [ObjectVision/GeoDMS#634]
So, for example, the brush colour can be edited even if a calculation rule configures it.
The edits are temporary and do not affect other open layers using the same classification.

![image](https://github.com/ObjectVision/GeoDMS/assets/48675272/26748538-e1ef-4782-9641-2ff82223c96a)

**Comment by eoudejans** [comment]
See issue #631

**Comment by MaartenHilferink** [comment]
Te implementeren in samenhang met #734

**Comment by MaartenHilferink** [comment]
Implemented as `a7475fac`, in samenhang met #734 as the comment above asked.

### What was wrong

`TreeItem::IsEditable()` is false as soon as an item carries a non-data-block calculation rule, and every edit path in the legend and the palette editor was gated on it. So the **Change Color** submenu was never built, the double-click and F2 paths fell through to ValueInfo, and Ramp reported *"Cannot change derived data; try to copy the attribute and change the copied data"* — a message that described exactly the manual workaround this issue asks to automate.

### What happens now

The first edit makes that copy. The values the rule currently computes are copied into `/Desktops/<desktop>/ViewData/<path of the item>/` under one epoch tick, the layer is re-themed onto the copy, and the edit is applied there. The configured item — shared with every other view that uses the same classification — is never written, and the event log names both items:

```
Edit: /Issue634/calc_classes/BrushColor is configured or calculated and is shared with every
view that uses it; /Desktops/Default/ViewData/Issue634/calc_classes/BrushColor was created
for this desktop and is edited instead (issue #634)
```

A second edit of the same cell reuses that copy, so it continues from the first edit rather than starting over from the configured values.

One mechanism covers all three kinds of cell: `PaletteControl` decides *what* may be copied — the class-break attribute, the label attribute and the per-class aspect palettes — and `DataItemColumn` routes `SetOrgColor`, `SetOrgText` and `Ramp` through that decision.

### What is deliberately not covered

- **The number of classes.** Adding or removing a class resizes the class unit itself rather than one of its attributes; that stays #734's Copy/Paste Classbreaks route, which generates a whole new classification for the view.
- **A rule that is a plain reference**, such as `PenColor := BrushColor`. `GetCurrSourceItem` resolves it to the referred item, which is writable, and that is where the edit has always landed. Nothing changes there.
- **Table views.** Without a `PaletteControl` owner nothing is copied, so a calculated attribute in a table view is refused exactly as before.
- The generated **Count / Area / SelCount** columns of a legend are calculated too, but an editable `pcount` is meaningless, so they are excluded by name.

### Verified

Release x64, full solution, against `MicroTst.dms` (new scripts `dmsscript/Issue634_layercontrol_cow.dmsscript` and `dmsscript/Issue634_tableview_negative.dmsscript` in GeoDMS-Test, `7f09ae8`):

- a calculated **colour**, **label** and **class break** each copied exactly once, and a repeat of the colour edit reported no second copy;
- the edits read back through #734's Copy Classbreaks as `2;rgb(255,255,255);"X"` against the configured `0;rgb(0,0,255);"calc class 0"`, with the other three classes untouched;
- the same edits in a plain table view: no copy, no write, no error;
- `MapViewClassification.dmsscript` and `Issue734_classbreak_copy_paste.dmsscript` still resolve every menu index and still produce their documented output, including the `Cannot paste class break -300 into class breaks of type uint32` refusal and the regenerate-on-configured-classification message.

`MicroTst` needed a new fixture for this: `hoek_classes/PenColor` and `SymbolColor` are plain references, so they resolve through to the editable `BrushColor` data block and never reach the new path. `/Issue634/calc_classes` has genuinely calculated class breaks, colours and labels.

### Fallout worth its own issue

Two things surfaced while testing, neither caused by this change:

1. `PaletteControl::CreateSymbolColumnFromLayer` collapses **every** colour aspect onto one theme slot, last-wins. On a layer that also carries `BrushColor`/`PenColor` parameters on its geometry, **Change Color** therefore edits the *pen* colour. Previously invisible, because the submenu was not built for a calculated palette at all; now it is. The `/Issue634` fixture works around it by giving the thematic attribute its own geometry.
2. `Copy Classbreaks` read the label attribute the *palette domain* resolves to rather than the one the view is showing, so after a label edit the legend and the clipboard disagreed. Fixed here as part of the change, since the two features have to agree on what "the labels of this classification" are.

Wiki: [[Map View Legend]] and [[Classification and Palette editor]].

**Comment by MaartenHilferink** [comment]
Follow-up to the fix above, on the test scripts rather than on the engine.

Writing the #634 regression test made it obvious that these scripts address pop-up menu items **by position**, and that this had already gone wrong. `MapViewClassification.dmsscript` documented index `{7}` as *"Copy Palette"* — an item that no longer exists — and had been firing *"Remove classBreaks"* instead; index `{1 2}` was commented *"Equal Count"* but is *"Equal Counts non-zero"*. Nothing failed, because a shifted index just fires whatever moved into that position. Revealing **Change Color** for a calculated palette, which is exactly what this issue does, shifts those indices again for anyone whose configuration has one.

So a menu path element may now be a quoted prefix of the item's caption (`10d24d63`):

```
SEND 4 2 1 "Edit Palette"
SEND 4 3 1 "Change Color" "Set to transparent"
```

Matching skips the `&` accelerator and is case-insensitive. The numeric form is unchanged on the wire, so scripts that keep using it send exactly the same bytes.

Three silent failures became errors, each reporting the captions that *are* available at that level: a name matching nothing, a name matching more than one item (which would otherwise pick the first), and an index past the end of its level. A sub-object index past the end of its object is an error for the same reason — it used to leave the container activated, so the rest of the script addressed the wrong control. And a **disabled** item is now refused rather than executed (`acd00bfb`): `MenuItem::Execute` never looked at `m_Flags`, so a script could fire something a user cannot pick.

Two things that surfaced while verifying, both worse than the numbering:

- **`MapViewClassification_CaptureMapView.dmsscript` was silently not running most of itself.** It waited 1100 ms after `DefaultView`, usually not enough for MicroTst to fetch its WMS tiles and populate the layer control. `WmCopyData` returns silently when there is no active `QDmsViewArea`, so every `SEND` after that was dropped with no log line, and the process still exited 0. Two consecutive runs stopped at different points. Both scripts now wait generously at the two moments where a view is being created — after `DefaultView` and after `Edit Palette`, which is made through `PostGuiOper` (`12a3914`, `06e848a`).
- That script now classifies with **JenksFisher non-zero**, the classification that runs on worker threads, so the captured map view exercises that path instead of the trivial one.

Test-repo commits: `f1acbcc`, `12a3914`, `06e848a` on top of `7f09ae8`.

Still open from the fix above, and not addressed here: `PaletteControl::CreateSymbolColumnFromLayer` collapses every colour aspect onto one theme slot, last-wins, so on a layer that also carries `BrushColor`/`PenColor` parameters on its geometry, **Change Color** edits the *pen* colour. That was invisible while the submenu was not built for a calculated palette at all. It deserves its own issue.
